### PR TITLE
[Core] Bring upstream updates from amgcl

### PR DESCRIPTION
**Pull request by ddemidov**
This is mostly to pull a fix to [the problem](https://groups.google.com/forum/#!topic/amgcl/AnV_l4MXkxE) reported in amgcl mailing list.

The relevant upstream commit is https://github.com/ddemidov/amgcl/commit/cfb806f57d8b0eb9a38216160cb0d1f83455e149.

### CODE REVIEW

File: external_libraries/amgcl/backend/builtin.hpp [hunk 1]
```diff
@@ -1029,11 +1029,41 @@ struct inner_product_impl<
...
+            s = t;
+        }
+
+        return s;
+    }
+
 #ifdef _OPENMP
-        return_type              _sum_stat[64];
+#  ifndef AMGCL_MAX_OPENMP_THREADS
+#    define AMGCL_MAX_OPENMP_THREADS 64
```
Comment (RiccardoRossi) [inline external_libraries/amgcl/backend/builtin.hpp:1061]: are u sure of this ? i guess you should check somewhere that this is enough

**Review by RiccardoRossi — APPROVED** [review]
other than that i approve

File: external_libraries/amgcl/backend/builtin.hpp (hunk 1 above)
Comment (ddemidov) [reply external_libraries/amgcl/backend/builtin.hpp:1061]: This was already set to 64. This change simply allows to customize the value with a preprocessor definition.

Comment (ddemidov) [reply external_libraries/amgcl/backend/builtin.hpp:1061]: Also this is not a hard limit, the inner product will work with larger number of cores, but will use more expensive dynamic vector in that case.

**Pull request merged**